### ISSUE TITLE
[Feature] Offers a way for doing `substreams gui --attach <endpoint>`

### ISSUE THREAD

**Issue by maoueh** [graphprotocol/graph-node#4826]
#### Background

We are actively working now on improving observability of a Substreams when consumed by a `sink` process. While doing this work inside the `substreams gui` command, we started to think on how Substreams actively being sinked using a `sink` process (`graph-node`, `substreams-sink-postgres`, `substreams-sink-kv`, etc.) could be "piped" to a GUI that would present all required observability.

We come  up with the idea that `sink` could offer a locally started gRPC server over a port that you could then "attach" to observe the Substreams progress. Idea would look like this:

```
substreams-sink-postgres run <dsn> <endpoint> <manifest> <module> --progress-port 7575
```

Then you would simply attach to this port:

```
substreams gui --attach localhost:7575
```

Internally the `substreams-sink-postgres` when receiving a gRPC connection on `7575` would simply forward received progress message to this external server, the `substreams gui --attach ...` would then receive those progress message and present it nicely, fully observing the stream progress.

The goal of this issue is to start a discussion of how things would look like in terms of `graph-node` first but then most broadly afterwards (studio, decentralized network, indexers, etc.)

#### First Step Idea

You first step would be to have something similar as what I've just described. `graph-node` would have a local gRPC server answering request.

Main difference, somehow a `deployment hash` would be passed when making the request (we are thinking of having a generic "id" field in the request where the hash could be put).

Then one could do something like:

```
substreams gui --attach "<host>:7575?id=Qm...."
```

Internally this would then find the right deployment, hook "somehow" into the right block stream and starts forwarding progress messages receives out to the gRPC connection.

##### Difficulties

We will need to find way to have some kind of "global" routing, the request would enters but then we would need to correctly hit the right node that is currently indexing the Substreams. Not sure what would be the blockers here, I know routing rules are in the config so a "single indexer" entity should be able to find out the right now to reach to.

After that, there is of course discussion to have about the Studio deployment an everything

##### Transport

A GraphQL subscription model could also be thought about instead of a gRPC connection. Using gRPC will make it easier to re-use the proto message directly and simply re-forward them to a gRPC connection. Also it makes it easier to implement the `--attach` thing on the `substreams gui` as a single protocol needs to be dealt with.

### Are you aware of any blockers that must be resolved before implementing this feature? If so, which? Link to any relevant GitHub issues.

Discussion(s)

### Some information to help us out

- [ ] Tick this box if you plan on implementing this feature yourself.
- [X] I have searched the issue tracker to make sure this issue is not a duplicate.

**Comment by matthewdarwin** [comment]
Instead of doing the logging this way, can we offer a way to the substreams tier1 to instead optionally forward messages to a "substreams reporting server".  This server can forward all the relevant messages that someone is interested.  It can also support a connect-web API so a web ui can built using the messages.

Then we don't need to change graph-node very much.

The workflow is instead: 
1. tier 1 connects to "substreams reporting server" and reports new trace ids at they become available (much like the authentication and metering plugins)
1. graph-node connects to substreams
1. upon connect, graph node stores the server and trace ID in a place accessible to the studio. This is all the information required to connect to the right "substreams reporting server" (the substreams reporting server URL might actually be recorded in the graph-node config)
1. if the substream is re-connected in graph-node, update this stored information
1. in studio when someone clicks "inspect", the web page generates a dialog box which connects to a web page, passing in the reporting server details
1. the web page contacts the reporting server and requests that the server send the stream of events
1. the reporting server tells the tier1 server to send the events for the requested trace id.
1. close down the event stream when the user closes the web page
1. done

**Comment by azf20** [comment]
Interesting @maoueh - confirming that the goal here is a "live" view of substreams progress (rather than anything more persistent)?

Not sure about the nicest way to combine this with the indexing status API (for example)

> We will need to find way to have some kind of "global" routing

I think this should be ok, Graph Node's config is generally aware of which index node is doing the work for a given subgraph.

**Comment by maoueh** [comment]
> Interesting @maoueh - confirming that the goal here is a "live" view of substreams progress (rather than anything more persistent)?

That's correct, this is a Substreams progress live view only, no persistence.

> I think this should be ok, Graph Node's config is generally aware of which index node is doing the work for a given subgraph.

That's my assumption also, but I'm unsure how exactly things work so there is uncertainties, for me at least.

> @matthewdarwin approach

I think your proposition is probably what we should aim for if possible, as this means global support on every sink(s). Requires more work however as the solution is a bit more involved, need some kind of global tracking on the indexer side, which means more operational cost (etcd?).

We could go faster with the in-sink approach while keeping the ultimate goal and ensuring we can perform a smooth transition. That is an in-between plan, in-sink than migrate to in-provider.
